packages/bsky/src/views/index.ts

```diff
@@ -1606,6 +1595,264 @@ export class Views {
+      const has3pBlock =
+        !state.ctx?.include3pBlocks && state.postBlocks?.get(uri)?.parent
+      if (has1pBlock || has3pBlock) {
+        return undefined
+      }
+      if (!this.viewerSeesNeedsReview({ uri, did: authorDid }, state)) {
+        return undefined
+      }
+
+      // No unauthenticated.
+      if (this.noUnauthenticatedPost(viewer, postView)) {
+        return undefined
+      }
+
+      // Hidden.
+      const { hiddenByThreadgate, mutedByViewer } = this.isHiddenThreadPost(
+        { rootUri, uri, viewer },
+        state,
+      )
+      // Is hidden reply.
+      if (hiddenByThreadgate || mutedByViewer) {
+        // Only show hidden anchor replies, not all hidden.
+        if (depth > 1) {
+          return undefined
+        }
+      } else if (depth === 1) {
+        // Don't include non-hidden anchor replies.
+        return undefined
+      }
+
+      // Recurse down.
+      const replies = this.threadHiddenV2Replies({
+        parentUri: uri,
+        viewer,
+        rootUri,
+        childrenByParentUri,
+        state,
+        below,
+        depth: depth + 1,
+      })
+
+      const item = this.threadHiddenV2ItemPost(
+        {
+          depth,
+          postView,
+          rootUri,
+          uri,
+          viewer,
+        },
+        state,
+      )
+
+      const tree: ThreadHiddenPostNode = {
+        item: item,
+        replies,
+      }
+
+      return tree
+    })
+  }
+
+  private threadHiddenV2ItemPostAnchor({
+    depth,
+    uri,
+  }: {
+    depth: number
+    uri: string
+  }): ThreadHiddenAnchorPostNode['item'] {
+    return {
+      $type: 'app.bsky.unspecced.getPostThreadHiddenV2#threadHiddenItem',
+      uri,
+      depth,
+      // In hidden replies, the anchor value is undefined, so it doesn't include the anchor in the result.
+      // This is helpful so we can use the same internal structure for hidden and non-hidden, while omitting anchor for hidden.
+      value: undefined,
+    }
+  }
+
+  private threadHiddenV2ItemPost(
+    {
+      depth,
+      postView,
+      rootUri,
+      uri,
+      viewer,
+    }: {
+      depth: number
+      postView: PostView
+      rootUri: string
+      uri: string
+      viewer: HydrateCtx['viewer']
+    },
+    state: HydrationState,
+  ): ThreadHiddenPostNode['item'] {
+    const { hiddenByThreadgate, mutedByViewer } = this.isHiddenThreadPost(
+      { rootUri, uri, viewer },
+      state,
+    )
+
+    const base = this.threadHiddenV2ItemPostAnchor({ depth, uri })
+    return {
+      ...base,
+      value: {
+        $type: 'app.bsky.unspecced.getPostThreadHiddenV2#threadHiddenItemPost',
+        post: postView,
+        hiddenByThreadgate,
+        mutedByViewer,
+      },
+    }
+  }
+
+  private isHiddenThreadPost(
+    {
+      rootUri,
+      uri,
+      viewer,
+    }: {
+      rootUri: string
+      uri: string
+      viewer: HydrateCtx['viewer']
+    },
+    state: HydrationState,
```

Comment:
Another example here where I believe we can pull `viewer` from `state.ctx`.  I think there are others we could look at but figured I'd just tag a couple to illustrate.